pkg/bff/users.go

```diff
@@ -54,121 +68,213 @@ func (s *Server) Login(c *gin.Context) {
+			c.JSON(http.StatusNotFound, api.ErrorResponse("organization not found"))
 			return
 		}
 
-		// Ensure the VASP record is correct for the user
-		if org.Testnet != nil && org.Testnet.Id != "" {
-			appdata.VASPs.TestNet = org.Testnet.Id
-		}
-		if org.Mainnet != nil && org.Mainnet.Id != "" {
-			appdata.VASPs.MainNet = org.Mainnet.Id
+		// Retrieve the user's collaborator record from the organization
+		var collaborator *models.Collaborator
+		if collaborator = org.GetCollaborator(*user.Email); collaborator == nil {
+			log.Debug().Str("email", *user.Email).Str("org_id", org.Id).Msg("could not find user in organization")
+			c.JSON(http.StatusUnauthorized, api.ErrorResponse("user is not authorized to access this organization"))
+			return
 		}
-	}
 
-	// Fetch user roles.
-	if roles, err = s.auth0.User.Roles(*user.ID); err != nil {
-		log.Error().Err(err).Msg("could not fetch roles associated with the user")
-		c.JSON(http.StatusInternalServerError, "could not complete user login")
-		return
+		// Other endpoints expect the user's verification status to be up to date
+		collaborator.Verified = *user.EmailVerified
 	}
 
-	if len(roles.Roles) == 0 {
-		// Assign the user the organization collaborator role
-		var role *management.Role
-		if role, err = s.FindRoleByName(CollaboratorRole); err != nil {
-			log.Error().Err(err).Msg("could not identify the default role to assign the user")
-			c.JSON(http.StatusInternalServerError, "could not complete user login")
-			return
+	if userRole == TSPRole {
+		// TSP users can be added to multiple organizations
+		appdata.AddOrganization(org.Id)
+	} else {
+		// Organizations with one collaborator need a leader to add other collaborators
+		if userRole == CollaboratorRole && len(org.Collaborators) == 1 {
+			userRole = LeaderRole
 		}
 
-		// Add the user as a collaborator in the organization record
-		collaborator := &models.Collaborator{
-			Email: *user.Email,
-		}
-		if err = org.AddCollaborator(collaborator); err != nil {
-			log.Error().Err(err).Msg("could not add the user to the organization collaborators")
-			c.JSON(http.StatusInternalServerError, "could not complete user login")
-			return
-		}
+		// Non-TSP users can only exist in one organization
+		if appdata.OrgID != "" && org.Id != appdata.OrgID {
+			// When switching to a new organization, make sure leader roles are not
+			// unintentionally preserved
+			if userRole == LeaderRole && len(org.Collaborators) > 1 {
+				userRole = CollaboratorRole
+			}
 
-		// Update the organization in the database
-		if err = s.db.UpdateOrganization(org); err != nil {
-			log.Error().Err(err).Msg("could not update the organization with the new collaborator")
-			c.JSON(http.StatusInternalServerError, "could not complete user login")
-			return
-		}
+			// Remove the collaborator record from the previous organization
+			// TODO: This might require a user confirmation prompt
```

Comment:
We will probably want a story for this - I think the idea is that this will be an additional endpoint to serve as confirmation.